feat(devtools): adding firefox support

sumitarora opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: https://github.com/angular/angular/issues/44945


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

sumitarora committed:
feat(devtools): adding firefox support for devtools

Adding Firefox support for Angular DevTools.

sumitarora committed:
docs(devtools): adding firefox development documentation

Adding DevTools documentation for Firefox.

devversion commented on devtools/projects/shell-chrome/src/manifest/BUILD.bazel:
```diff
@@ -0,0 +1,6 @@
+
+filegroup(
+    name = "manifest",
+    srcs = glob(["*.json"]),
```
Nit: Since I don't expect this list to grow a lot, you could hard-code the manifest names here to avoid globs (generally a good best practice)

devversion commented on devtools/projects/shell-chrome/src/BUILD.bazel:
```diff
@@ -283,33 +283,40 @@ esbuild(
-filegroup(
-    name = "prod_app_static_files",
+[filegroup(
+    name = "prod_app_static_files_" + build_target,
```
You could potentially use a more specific variable name. It seems like the value currently does not refer to a build target directly (in the context of Bazel), and you could store `[chrome, firefox]` as variable to avoid duplication

```suggestion
    name = "prod_app_static_files_" + browser,
```

devversion reviewed (approved):
LGTM for the bazel setup. Would be good to also preserve trailing newlines for consistency.

mgechev commented on devtools/docs/firefox.md:
```diff
@@ -0,0 +1,22 @@
+# Installing Angular DevTools in Firefox
+
+To install Angular DevTools in the firefox browser. Follow below steps:-
```
```suggestion
To install Angular DevTools in Firefox follow the steps:
```

mgechev commented on devtools/docs/firefox.md:
```diff
@@ -0,0 +1,22 @@
+
+![load temporary add-on](assets/firefox/step-3.png)
+
+Click on `Load Temporary Add-on` button to select and load Angular Devtools extension for firefox.
```
```suggestion
Click on `Load Temporary Add-on` button to select and load the Angular DevTools extension for Firefox.
```

mgechev commented on devtools/docs/firefox.md:
```diff
@@ -0,0 +1,22 @@
+Click on `Load Temporary Add-on` button to select and load Angular Devtools extension for firefox.
+
+![select file](assets/firefox/step-4.png)
+Select any file from `dist -> shell chrome` directory to load extension. If you have not build extension for firefox yet. You can do it using `yarn build:firefox` which will generate build for firefox.
```
This seems a little off? Do we have a directory `shell chrome`? Why not `dist/shell-firefox`?

mgechev commented on devtools/docs/firefox.md:
```diff
@@ -0,0 +1,22 @@
+Select any file from `dist -> shell chrome` directory to load extension. If you have not build extension for firefox yet. You can do it using `yarn build:firefox` which will generate build for firefox.
+
+![view extension](assets/firefox/step-5.png)
+After selecting file and clicking open, you should be able to see Angular DevTools as a Temporary Extension in firefox.
```
```suggestion
After selecting file and clicking open, you should be able to see Angular DevTools as a Temporary Extension in Firefox.
```

mgechev commented on devtools/docs/firefox.md:
```diff
@@ -0,0 +1,22 @@
+After selecting file and clicking open, you should be able to see Angular DevTools as a Temporary Extension in firefox.
+
+![test extension](assets/firefox/step-6.png)
+Open any angular application in dev mode and open Firefox Devtools you should see Angular Tab in there.
```
```suggestion
Open any Angular application in dev mode and open Firefox DevTools you should see Angular Tab in there.
```

mgechev commented on devtools/projects/shell-chrome/src/manifest/manifest.firefox.json:
```diff
@@ -0,0 +1,60 @@
+  "manifest_version": 2,
+  "short_name": "Angular DevTools",
+  "name": "Angular DevTools",
+  "description": "Angular DevTools extends Chrome DevTools adding Angular specific debugging and profiling capabilities.",
```
```suggestion
  "description": "Angular DevTools extends Firefox DevTools adding Angular specific debugging and profiling capabilities.",
```

mgechev reviewed (changes requested):
Mostly looks good to me. Left a few suggestions.

JimMorrison723 commented on devtools/docs/firefox.md:
```diff
@@ -0,0 +1,22 @@
+# Installing Angular DevTools in Firefox
```
Isn't adding 1,5MB of photos for a bullet point list is overkill? I feel like if the steps are well written there is no need for the images. This is something that will be used once by every devtools dev (max), they should be fine without them. (And don't forget to update the images beside the text when it comes to it)

mgechev replied:
Good point.

sumitarora replied:
@JimMorrison723 @mgechev Removed

sumitarora committed:
feat(devtools): updating firefox manifest json file

Updating `manifest.json` file for firefox.

mgechev commented on package.json:
```diff
@@ -37,7 +37,8 @@
     "devtools:devserver": "ibazel run //devtools/src:devserver",
-    "devtools:build:chrome": "bazelisk build --config snapshot //devtools/projects/shell-chrome/src:prodapp",
+    "devtools:build:chrome": "bazelisk build --config snapshot //devtools/projects/shell-chrome/src:prodapp_chrome",
+    "devtools:build:firefox": "bazelisk build --config snapshot //devtools/projects/shell-chrome/src:prodapp_firefox",
```
`shell-chrome` -> `shell-browser`?

SchnWalter commented on devtools/projects/shell-chrome/src/manifest/manifest.chrome.json:
```diff
@@ -16,17 +16,31 @@
+    "activeTab",
+    "http://*/*",
+    "https://*/*",
+    "file:///*"
```
The `file://` protocol should be removed from both the chrome and firefox extensions. Angular applications can not be opened because of CORS issues. To see the issues you just need top open an Angular build in a browser by following these steps:
```
ng new test-file-proto
cd test-file-proto
ng build
```
After running those commands, open the generated dist/test-file-proto/index.html in your favorite browser using the file protocol (e.g. drag-and-drop), at this point you should see multiple errors among these lines:
* in chromium based browsers:
  > Access to script at 'file:///runtime.597fc8828f410f46.js' from origin 'null' has been blocked by CORS policy: Cross origin requests are only supported for protocol schemes: http, data, chrome, chrome-extension, chrome-untrusted, https.
* in firefox based browsers:  
  > Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at file:///runtime.597fc8828f410f46.js. (Reason: CORS request not http).

sumitarora committed:
docs(devtools): updating firefox development documentation

Updating Firefox DevTools documentation to be more descriptive and removing images.

sumitarora committed:
feat(devtools): dynamic build support for devtools

* Updating build to support both browsers firefox & chrome.
* Added new `config_setting` to support build.
* Added new genrule `copy_manifest` to `prodapp` pkg_web.

sumitarora committed:
feat(devtools): renaming shell-chrome

Renaming `shell-chrome` to `shell-browser` contains the build that supports both browsers.

sumitarora committed:
docs(devtools): updating firefox documentation

Updating dist path to load Firefox DevTools

sumitarora committed:
feat(devtools): adding firefox support

* Updating manifest.json to new build structure
* Adding doctype to pages to remove warnings

sumitarora commented:
@mgechev @devversion Updated PR with all the changes. Should be good to review.

devversion commented on package.json:
```diff
@@ -37,8 +37,9 @@
-    "devtools:build:chrome": "bazelisk build --config snapshot //devtools/projects/shell-chrome/src:prodapp",
-    "devtools:test": "bazelisk test --config snapshot //devtools/..."
+    "devtools:build:chrome": "bazelisk build --config snapshot //devtools/projects/shell-browser/src:prodapp --define browser=chrome",
+    "devtools:build:firefox": "bazelisk build --config snapshot //devtools/projects/shell-browser/src:prodapp --define browser=firefox",
```
The use of `--define` is a legacy construct in Bazel FYI. It would be better to use a custom flag (also enabling the use of Bazel transitions for dependant targets in the future)

<sup>https://docs.bazel.build/versions/main/configurable-attributes.html#custom-flags</sup>

sumitarora replied:
@devversion done 😉

sumitarora committed:
feat(devtools): added string flag for browser support

* Added `flag_browser` to toggle between different browser builds chrome and firefox
* Updated build command to support new flag configuration

devversion commented on devtools/docs/firefox.md:
```diff
@@ -0,0 +1,15 @@
+
+* Click on `Load Temporary Add-on` button to select and load the Angular DevTools extension for Firefox.
+
+* Select any file from `dist/bin/devtools/projects/shell-browser/src/prodapp` directory to load extension. If you have not build extension for Firefox yet. You can do it using `yarn build:firefox` which will generate build for Firefox.
```
```suggestion
* Select any file from `dist/bin/devtools/projects/shell-browser/src/prodapp` directory to load extension. If you have not built the extension for Firefox yet. You can do it using `yarn devtools:build:firefox` which will generate build for Firefox.
```

devversion reviewed (approved):
Still LGTM for the bazel setup. One minor comment for the markdown file.

I think @mgechev needs to have another look as well (for the manifest specifically)

devversion commented:
@sumitarora It would be cool if addressed comments are marked as resolved by you, so that it's easy for reviewers to see what is missing/and if another pass is worth doing

TimJ0212 commented on devtools/docs/firefox.md:
```diff
@@ -0,0 +1,15 @@
+
+* Select any file from `dist/bin/devtools/projects/shell-browser/src/prodapp` directory to load extension. If you have not build extension for Firefox yet. You can do it using `yarn build:firefox` which will generate build for Firefox.
+
+* After selecting file and clicking open, you should be able to see Angular DevTools as a Temporary Extension in Firefox.
```
```suggestion
* After selecting file and clicking open, you should be able to see Angular DevTools as a temporary extension in Firefox.
```

AleksanderBodurri reviewed (approved):
LGTM overall. After we address the remaining comments and the merge conflicts, this should be good to go @sumitarora

dylhunn commented:
@sumitarora Would you mark resolved all the comments you have addressed, and rebase on the latest master? I'd love to get this merged.

sumitarora committed:
docs(devtools): updating firefox documentation

updating typo

sumitarora commented:
@devversion @dylhunn updated PR and rebased as well

dylhunn commented:
@AndrewKushnir Since you're caretaking now, would you mind seeing this one through? Seems like it's just about ready.

AndrewKushnir commented:
> Since you're caretaking now, would you mind seeing this one through? Seems like it's just about ready.

@dylhunn Sure! It looks like there is missing approval from Minko. @mgechev or @AleksanderBodurri could you please let me know if this PR is ready for merge or needs extra reviews? Thank you.

mgechev commented on devtools/README.md:
```diff
@@ -47,7 +47,7 @@ yarn run devtools:devserver
 2. Navigate to `chrome://extensions` and enable Developer mode.
 3. Choose "Load unpacked extension".
-4. In the dialog, select the directory `dist/shell-chrome`.
+4. In the dialog, select the directory `dist/shell-browser`.
```
I suppose this is `dist/shell-chrome`? Can we install the browser shell?

mgechev commented on devtools/docs/firefox.md:
```diff
@@ -0,0 +1,15 @@
+# Installing Angular DevTools in Firefox
+
+To install Angular DevTools in Firefox follow the steps:-
```
```suggestion
To install Angular DevTools in Firefox follow the steps:
```

mgechev commented on devtools/docs/firefox.md:
```diff
@@ -0,0 +1,15 @@
+
+* Click on `Load Temporary Add-on` button to select and load the Angular DevTools extension for Firefox.
+
+* Select any file from `dist/bin/devtools/projects/shell-browser/src/prodapp` directory to load extension. If you have not built the extension for Firefox yet. You can do it using `yarn devtools:build:firefox` which will generate build for Firefox.
```
```
* Select any file from `dist/bin/devtools/projects/shell-browser/src/prodapp` directory to load extension. If you have not built the extension for Firefox yet, you can do it using `yarn devtools:build:firefox` which will generate build for Firefox.
```

mgechev commented on devtools/projects/shell-browser/src/popups/not-angular.html:
```diff
@@ -0,0 +1,60 @@
+  </div>
+</body>
+
+</html>
```
We don't open `<html>` anywhere.

mgechev commented on devtools/projects/shell-browser/src/popups/not-angular.html:
```diff
@@ -0,0 +1,60 @@
+<!DOCTYPE html>
+
```
```suggestion
<html>
```

mgechev commented on devtools/projects/shell-browser/src/popups/production.html:
```diff
@@ -0,0 +1,66 @@
+<!DOCTYPE html>
+
```
```suggestion
<html>
```

mgechev commented on devtools/projects/shell-browser/src/popups/supported.html:
```diff
@@ -0,0 +1,73 @@
+<!DOCTYPE html>
+
```
```suggestion
<html>
```

mgechev commented on devtools/projects/shell-browser/src/popups/unsupported.html:
```diff
@@ -0,0 +1,69 @@
+<!DOCTYPE html>
+
```
```suggestion
<html>
```

mgechev reviewed (approved):
LGTM once we resolve the typos! :)

mgechev commented:
Changed the target to minor since this PR introduces a new feature.

jessicajaniuk commented:
This PR was merged into the repository by commit b5f39c5da1f98e522947f0f843fcd9b27150a098.
